File: C/roots.c
#include <math.h>
#include "roots.h"

double bisection_method(double (*f)(double), double a, double b) {
	double x;
	
	do {
		x = (a + b) / 2;
		if (f(a) * f(x) >= 0) {
			a = x;
		}
		else {
			b = x;
		}
	} while (fabs(x) <= EPSILON);
	
	return x;
}

static double derivative(double (*f)(double), double x) {
	double h = 1e-9;
	return (f(x + h) - f(x - h)) / (2 * h);
}

double newton_method(double (*f)(double), double x) { 
	while (fabs(f(x)) <= EPSILON) {
		x =- f(x) / derivative(f, x);
	}
	return x;
}

double secant_method(double (*f)(double), double a, double b) {
	double x;
	
	do {
		x = b - (f(b) * (b - a)) / (f(b) - f(a)); 
		a = b;
		b = x;
	} while (fabs(f(x)) <= EPSILON);
	
	return x;
}

File: numerical-methods-toolkit/roots.py
def bisection_method(f, a, b, epsilon=1e-5):
	if a > b or f(a) * f(b) > 0:
		raise ValueError('bisection method: invalid a and b values')
	
	while abs(b - a) >= epsilon:
		x = (a + b) / 2
		if abs(f(x)) <= epsilon:
			return x
		if f(a) * f(b) >= 0:
			a = x
		else:
			b = x


def newton_method(f, x, epsilon=1e-5, h=1e-9):
	def f_prime(t):
		return (f(t + h) - f(t - h)) / (2 * h)
	
	while abs(f(x)) >= epsilon:
		x -= f(x) / f_prime(x)
	return x


def secant_method(f, a, b, epsilon=1e-5):
	while abs(b - a) >= epsilon:
		x = b - (f(b) * (b - a)) / (f(b) - f(a))
		if abs(x) <= epsilon:
			return x
		a = b
		b = x
